Sources/NIOSSL/SSLContext.swift

```diff
@@ -779,6 +791,90 @@ extension NIOSSLContext {
+                // If we do and we are executing this callback then the future has already resolved
+                // We can safely call wait() which will resolve immediately
+                let userChosenContext: NIOSSLContext
+                do {
+                    userChosenContext = try futureSSLContext.wait()
+                } catch {
+                    return SSL_TLSEXT_ERR_NOACK
+                }
+
+                // The user may or may not retain a reference to the NIOSSLContext
+                // that they return here, if they provide one. The NIOSSLContext finalizer
+                // explicitly decrements the reference count on the underlying SSL_CTX.
+                //
+                // However, SSL_set_SSL_CTX explicitly increases the reference count and
+                // will decrement the count when the SSL* object is destroyed, thus finally
+                // freeing the raw SSL_CTX object, even if the swift reference is orphaned.
+                //
+                // As such, no memory should be leaked or freed too early in the event that
+                // the user fails to retain their created and return NIOSSLContext. In the
+                // event that they do retain it, the reference count will remain at at-least
+                // 1 and thus they can re-use the same context for multiple connections.
+                let nativeContextValue = withExtendedLifetime(userChosenContext) {
+                    return CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext)
+                }
+
+                // CNIOBoringSSL_SSL_set_SSL_CTX should return a pointer to the very same context it was
+                // told to assign on success. On error, this will not happen, so we check for equality
+                // here and handle any such errors by signaling the connection to fail.
+                if nativeContextValue != userChosenContext.sslContext {
+                    return SSL_TLSEXT_ERR_NOACK
+                }
+
+                return SSL_TLSEXT_ERR_OK
+            }
+
+            // Make sure we get the server name extracted.
+            guard let cServerName = CNIOBoringSSL_SSL_get_servername(ssl, CNIOBoringSSL.TLSEXT_NAMETYPE_host_name) else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            let serverName = String(cString: cServerName)
+
+            // Can do nothing about empty indicators.
+            guard !serverName.isEmpty else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Construct extension values
+            let values = NIOSSLClientExtensionValues(serverName: serverName)
+
+            // Issue call to get a new ssl context
+            let futureSSLContext: EventLoopFuture<NIOSSLContext>
+            do {
+                futureSSLContext = try parentSwiftContext.sslContextCallback!(values, parentSwiftContext)
+            } catch {
+                return SSL_TLSEXT_ERR_NOACK
```

Comment:
This is a complex ask, but it would be my preference that we didn't throw the user's error away. Ideally we'd persist the error and propagate it down the pipeline, which will require us to store it somewhere on the `SSLContext` object.

Reply:
My new approach uses a saved result on the `SSLContext` so I decided to save the error as a failed result. This feels right, as the synchronous error here is basically the same as the future resolving to an error later. We would want to handle those types of errors in same way. Let me know if that makes sense, and where we can propagate further.